src/Distributed/distributed_incompressible_model.jl

```diff
@@ -0,0 +1,97 @@
+using MPI
+
+using Oceananigans
+using Oceananigans.Grids
+using Oceananigans.Grids: halo_size
+
+#####
+##### Distributed model struct and constructor
+#####
+
+struct DistributedIncompressibleModel{A, G, M}
+    architecture :: A
+            grid :: G
+           model :: M
+end
+
+function DistributedIncompressibleModel(; architecture, grid, boundary_conditions=nothing, model_kwargs...)
+    my_rank = architecture.my_rank
+    i, j, k = architecture.my_index
+    Rx, Ry, Rz = architecture.ranks
+    my_connectivity = architecture.connectivity
+
+    ## Construct local grid
+
+    Nx, Ny, Nz = size(grid)
+
+    # Pull out left and right endpoints for full model.
+    xL, xR = grid.xF[1], grid.xF[Nx+1]
+    yL, yR = grid.yF[1], grid.yF[Ny+1]
+    zL, zR = grid.zF[1], grid.zF[Nz+1]
+    Lx, Ly, Lz = length(grid)
+
+    # Make sure we can put an integer number of grid points in each rank.
+    @assert isinteger(Nx / Rx)
+    @assert isinteger(Ny / Ry)
+    @assert isinteger(Nz / Rz)
+
+    nx, ny, nz = Nx÷Rx, Ny÷Ry, Nz÷Rz
+    lx, ly, lz = Lx/Rx, Ly/Ry, Lz/Rz
+
+    x₁, x₂ = xL + (i-1)*lx, xL + i*lx
+    y₁, y₂ = yL + (j-1)*ly, yL + j*ly
+    z₁, z₂ = zL + (k-1)*lz, zL + k*lz
+
+    # FIXME: local grid might have different topology!
```

Comment:
Local grids would never be periodic in the direction we do the split.  Would they always be bounded in that direction?

Reply:
Yes I think so. I guess I wasn't as sure when first writing the code. For time-stepping I don't think it matters right now since `HaloCommunication` boundary conditions are injected in the right places no matter the topology, but it would be good to get the topologies right soon.

Reply:
It runs in this case so I'm happy to move forward and then generallize this sometime later